File: javascript/getting-started/server/task.js
import { mongoose } from "mongoose";
import { TaskModel } from "./models/task";
import { GenezioDeploy } from "@genezio/types";

const red_color = "\x1b[31m%s\x1b[0m";
const missing_env_error =
  "ERROR: Your MONGO_DB_URI environment variable is not properly set, go to https://genez.io/blog/how-to-add-a-mongodb-to-your-genezio-project/ to learn how to integrate your project with Mongo DB";

/**
 * The Task server class that will be deployed on the genezio infrastructure.
 */
@GenezioDeploy()
export class TaskService {
  constructor() {
    mongoose.set("strictQuery", true);
    this.#connect();
  }

  /**
   * Private method used to connect to the DB.
   */
  #connect() {
    if (!process.env.MONGO_DB_URI) {
      console.log(red_color, missing_env_error);
      return;
    }
    mongoose.connect(process.env.MONGO_DB_URI || "").catch((err) => {
      console.log(err);
      throw err;
    });
  }

  /**
   * Method that returns all tasks for a giving user ID.
   * Only authenticated users with a valid token can access this method.
   
   * The method will be exported via SDK using genezio.
   *
   * @param {*} token The user's token.
   * @param {*} userId The user ID
   * @returns An object containing two properties: { success: true, tasks: tasks }
   */
  async getAllTasksByUser(token) {
    if (!process.env.MONGO_DB_URI) {
      console.log(red_color, missing_env_error);
      return { success: false, err: missing_env_error };
    }
    console.log(`Get all tasks by user request received with token ${token}`);

    let tasks;
    try {
      tasks = (await TaskModel.find({ token: token })).map((task) => {
        return {
          id: task._id.toString(),
          title: task.title,
          solved: task.solved,
          token: task.token,
        };
      });
    } catch (error) {
      return { success: false, err: error };
    }

    if (tasks.length === 0) {
      try {
        await TaskModel.create({
          token: token,
          title: "Check the other example projects",
          url: "https://github.com/Genez-io/genezio-examples",
        });

        await TaskModel.create({
          token: token,
          title: "Check our documentation",
          url: "https://docs.genez.io/genezio-documentation/",
        });

        await TaskModel.create({
          token: token,
          title: "Watch our Youtube tutorials",
          url: "https://www.youtube.com/@genezio7235",
        });

        await TaskModel.create({
          token: token,
          title: "Read our technical articles on genezio blog",
          url: "https://genez.io/blog/",
        });

        const initTasks = (await TaskModel.find({ token: token })).map(
          (task) => {
            return {
              id: task._id.toString(),
              title: task.title,
              solved: task.solved,
              token: task.token,
            };
          }
        );

        return { success: true, tasks: initTasks };
      } catch (error) {
        return { success: false, err: err };
      }
    }

    return { success: true, tasks: tasks };
  }

  /**
   * Method that creates a task for a giving user ID.
   * Only authenticated users with a valid token can access this method.
   *
   * The method will be exported via SDK using genezio.
   *
   * @param {*} token The user's token.
   * @param {*} title The tasktitle.
   * @param {*} ownerId The owner's of the task ID.
   * @returns An object containing two properties: { success: true, tasks: tasks }
   */
  async createTask(token, title) {
    if (!process.env.MONGO_DB_URI) {
      console.log(red_color, missing_env_error);
      return { success: false, err: missing_env_error };
    }
    console.log(
      `Create task request received for user with ${token} with title ${title}`
    );

    let task;
    try {
      task = await TaskModel.create({
        title: title,
        token: token,
      });
    } catch (error) {
      return { success: false, err: error };
    }

    return {
      success: true,
      task: { title: title, token: token.toString(), id: task._id.toString() },
    };
  }

  /**
   * Method that creates a task for a giving user ID.
   * Only authenticated users with a valid token can access this method.
   *
   * The method will be exported via SDK using genezio.
   *
   * @param {*} token The user's token.
   * @param {*} id The task's id.
   * @param {*} title The task's title.
   * @param {*} solved If the task is solved or not.
   * @returns An object containing two properties: { success: true }
   */
  async updateTask(token, id, title, solved) {
    if (!process.env.MONGO_DB_URI) {
      console.log(red_color, missing_env_error);
      return { success: false, err: missing_env_error };
    }
    console.log(
      `Update task request received with id ${id} with title ${title} and solved value ${solved}`
    );

    try {
      await TaskModel.updateOne(
        { _id: id, token: token },
        {
          title: title,
          solved: solved,
        }
      );
    } catch (error) {
      return { success: false, err: error };
    }

    return { success: true };
  }

  /**
   * Method that deletes a task for a giving user ID.
   * Only authenticated users with a valid token can access this method.
   *
   * The method will be exported via SDK using genezio.
   *
   * @param {*} token The user's token.
   * @param {*} title The tasktitle.
   * @param {*} ownerId The owner's of the task ID.
   * @returns An object containing one property: { success: true }
   */
  async deleteTask(token, id) {
    if (!process.env.MONGO_DB_URI) {
      console.log(red_color, missing_env_error);
      return { success: false, err: missing_env_error };
    }
    console.log(`Delete task with id ${id} request received`);

    try {
      await TaskModel.deleteOne({ token: token, _id: id });
    } catch (error) {
      return { success: false, err: error };
    }

    return { success: true };
  }
}

File: dart/getting-started/server/lib/task.dart
import 'package:mongo_dart/mongo_dart.dart';
import 'package:getting_started/helper.dart';
import 'package:json_annotation/json_annotation.dart';

part 'task.g.dart';

@JsonSerializable()
class Task {
  String id;
  String title;
  String url;
  String token;
  bool solved;
  DateTime date;

  Task(
    this.id,
    this.title,
    this.url,
    this.token,
    this.solved,
    this.date,
  );

  factory Task.fromJson(Map<String, dynamic> json) => _$TaskFromJson(json);

  Map<String, dynamic> toJson() => _$TaskToJson(this);
}

@JsonSerializable()
class GetTasksResponse {
  bool success;
  List<Task> tasks;

  GetTasksResponse(
    this.success,
    this.tasks,
  );

  factory GetTasksResponse.fromJson(Map<String, dynamic> json) =>
      _$GetTasksResponseFromJson(json);

  Map<String, dynamic> toJson() => _$GetTasksResponseToJson(this);
}

@JsonSerializable()
class GetTaskResponse {
  bool success;
  Task task;

  GetTaskResponse(
    this.success,
    this.task,
  );

  factory GetTaskResponse.fromJson(Map<String, dynamic> json) =>
      _$GetTaskResponseFromJson(json);

  Map<String, dynamic> toJson() => _$GetTaskResponseToJson(this);
}

@JsonSerializable()
class UpdateTaskResponse {
  bool success;

  UpdateTaskResponse(
    this.success,
  );

  factory UpdateTaskResponse.fromJson(Map<String, dynamic> json) =>
      _$UpdateTaskResponseFromJson(json);

  Map<String, dynamic> toJson() => _$UpdateTaskResponseToJson(this);
}

@JsonSerializable()
class DeleteTaskResponse {
  bool success;

  DeleteTaskResponse(
    this.success,
  );

  factory DeleteTaskResponse.fromJson(Map<String, dynamic> json) =>
      _$DeleteTaskResponseFromJson(json);

  Map<String, dynamic> toJson() => _$DeleteTaskResponseToJson(this);
}

class TaskService {
  Db? db;

  /// Method that returns all tasks for a given user ID.
  ///
  /// The method will be exported to a SDK using genezio.
  ///
  /// @param token The user's token.
  /// @returns A list containing tasks.
  Future<GetTasksResponse> getAllTasksByUser(String token) async {
    print("Trying to get all tasks by user with token: $token");

    // Check if the database is connected
    if (db == null) {
      // Connect to the database
      await _connect();
    }

    // Get tasks from database
    final tasks = await db
        ?.collection(TASK_COLLECTION)
        .find({
          "token": token,
        })
        .map((task) {
          return Task.fromJson(task);
        })
        .toList()
        .catchError((e) {
          print("Error getting tasks from database: $e");
          throw e;
        });

    if (tasks == null || tasks.isEmpty) {
      await db
          ?.collection(TASK_COLLECTION)
          .insert(Task(
                  ObjectId().$oid,
                  "Check our documentation",
                  "https://docs.genez.io/genezio-documentation/",
                  token,
                  false,
                  DateTime.now())
              .toJson())
          .catchError((e) {
        print("Error adding task to database: $e");
        throw e;
      });

      await db
          ?.collection(TASK_COLLECTION)
          .insert(Task(
                  ObjectId().$oid,
                  "Watch our Youtube tutorials",
                  "https://www.youtube.com/@genezio7235",
                  token,
                  false,
                  DateTime.now())
              .toJson())
          .catchError((e) {
        print("Error adding task to database: $e");
        throw e;
      });

      await db
          ?.collection(TASK_COLLECTION)
          .insert(Task(
                  ObjectId().$oid,
                  "Read our technical articles on genezio blog",
                  "https://genez.io/blog",
                  token,
                  false,
                  DateTime.now())
              .toJson())
          .catchError((e) {
        print("Error adding task to database: $e");
        throw e;
      });
    } else {
      return GetTasksResponse(true, tasks);
    }

    // Get tasks from database
    final initTasks = await db
        ?.collection(TASK_COLLECTION)
        .find({
          "token": token,
        })
        .map((task) {
          return Task.fromJson(task);
        })
        .toList()
        .catchError((e) {
          print("Error getting tasks from database: $e");
          throw e;
        });

    // Check if iniTasks is null
    if (initTasks == null || initTasks.isEmpty) {
      return GetTasksResponse(false, []);
    }

    return GetTasksResponse(true, initTasks);
  }

  /// Method that creates a task for a giving user ID.
  ///
  /// @param token The user's token.
  /// @returns A list the string "success" if everything went well.
  Future<GetTaskResponse> createTask(
      String token, String title, String url) async {
    print("Trying to add a new task with title: $title");

    // Check if the database is connected
    if (db == null) {
      // Connect to the database
      await _connect();
    }

    // Create a new task in the database
    Task task = Task(ObjectId().$oid, title, url, token, false, DateTime.now());

    // Add task into the database
    await db?.collection(TASK_COLLECTION).insert(task.toJson()).catchError((e) {
      print("Error adding task to database: $e");
      throw e;
    });

    return GetTaskResponse(true, task);
  }

  /// Method that deletes a task for giving user token and with a given id.
  ///
  /// @param token The user's token.
  /// @param id The tasks's id.
  /// @returns A list the string "success" if everything went well.
  Future<DeleteTaskResponse> deleteTask(String token, String id) async {
    print("Trying to delete the task with id: $id");

    // Check if the database is connected
    if (db == null) {
      // Connect to the database
      await _connect();
    }

    // Delete the task from the database
    await db?.collection(TASK_COLLECTION).remove({
      "id": id,
      "token": token,
    }).catchError((e) {
      print("Error deleting task from database: $e");
      throw e;
    });

    return DeleteTaskResponse(true);
  }

  /// Method that updates a task.
  ///
  /// @param token The user's token.
  /// @param id The tasks's id.
  /// @param title The tasks's title.
  /// @param solved The tasks's solved flag.
  /// @returns A list the string "success" if everything went well.
  Future<UpdateTaskResponse> updateTask(
      String id, String token, String title, String url, bool solved) async {
    print("Trying to update the task with id: $id");

    // Check if the database is connected
    if (db == null) {
      // Connect to the database
      await _connect();
    }

    // Update the task in the database
    await db?.collection(TASK_COLLECTION).update({
      "id": id,
      "token": token,
    }, {
      r"$set": {
        "title": title,
        "url": url,
        "solved": solved,
      }
    }).catchError((e) {
      print("Error updating task in database: $e");
      throw e;
    });

    return UpdateTaskResponse(true);
  }

  /// Method that connects to the database.
  ///
  /// @returns A string "success" if everything went well.
  Future<String> _connect() async {
    // Connect to the database
    db = await Db.create(MONGODB_URI).catchError((e) {
      print("Error connecting to database: $e");
      throw e;
    });

    // Open the database
    await db?.open().catchError((e) {
      print("Error opening database: $e");
      throw e;
    });

    return "success";
  }
}
